WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+    // first finding transitions, and then finding the smallest one.
+
+    // I wish I knew spark well enough already to be able to split the RDD of Hits in partitions
+    // which do not break a transition across partitions, in order to avoid a synchronization step.
+    // I immediately see two approaches that would work: 
+    // 1) If I could include the hit which happens to be part of a signal transition in both partitions, I'd be sure
+    // that no transition is broken up across partitions, and then the partitions 
+    // would only need to communicate back their winning candidate. seems fail-safe
+    // 2) If I could align the partition boundaries to only fall in the middle of hits of the same word
+    // we also would not have a broken transition.
+    // But if we only do option 2, our number of partitions would depend strongly on the input text.
+    
+    // Without a solution for avoiding the synchronization, I don't yet have a sense for how much of an improvement 
+    // it gives to do this part of the computation on the cluster instead of in the driver.
+    // For a non-pathological case, the filtering all words to just our word1 and word2 should already be 
+    // a huge shrink in dataset size.
+    
+    // but since there was a big warning against the trivial "make it run on spark" 
+    // implementation of immediately collecting, I'll try to write it in a spark-compatible style.
+    // 
+    // Stylewise, I'm unhappy because it becomes a bit car/cdr soup though, 
+    // which could be solved by defining case classes.
+
+    // I can't find a spark foldLeft which lets me fold Hits into SameWordIntervals.
+    // A verbose workaround but pretty foolproof solution would be to create a trait as union type.
+    // In this case I took the "wrap everything as a list" approach. Pretty ugly.
+    // I'm not sure whether defining a custom accumulator would help for making it less ugly 
+    // and for making it more performant.
+     
+    val reverseOrderedSameWordIntervals:List[SameWordInterval] = convertHitsToSameWordIntervals(orderedHits)
+    // the fact that I do a fold which is an action instead of a transform, 
+    // is bumping me out of RDDs into regular scala-land at this point anyway.
+    // At this point we have had another huge data shrink from skipping consecutive hits of the same word,
+    // and it is already late, so I'm going to do the last part in pure scala
+    
+    // println("""***** """+reverseOrderedSameWordIntervals.take(100).toList.mkString(", "))
+    
+    // if one of the words does not exist, there will only be one element in this list
+    if (reverseOrderedSameWordIntervals.size == 1) {
+      None 
+    } else {
+      Some(findMinimumDistanceBetweenAlternatingIntervals(reverseOrderedSameWordIntervals))
+    }
+  }
+   
+  def convertHitsToSameWordIntervals(orderedHits:RDD[Hit]):List[SameWordInterval] = {
```

Comment:
Is it clear after the explanation higher up in the first note, of why I bother with this data size reduction at all?
Or do you want me to try explaining it here differently?